### PR TITLE
Aditya | Week3 - Implementation of assignment 3 and routing

### PR DESCRIPTION
**Pull request by aditya-s-19**
Pull Request Questionnaire

(1) Briefly describe your change.
I have integrated my week-3 assignment, implemented routing, did basic testing and updated also according to all the feedback 

(2) What parts of the application are affected by your change?
Affects my folder but also Keshav's because I had to save them all with ESlint settings as they were giving me error for eol.

(3) How did you test your change? If relevant, which role(s) and/or customer configuration(s) did you use?
I tested by trying all corner cases fixing them.

Root Cause (if any)?
-

### CODE REVIEW

**Comment by anusreesubash** [comment]
Please rebase to main and resolve conflicts to not have other changes in you PR.

File: interns-react-weekly-assignments/.eslintrc.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
     ],
     "rules": {
         // "no-duplicate-imports": "error"
+        "eol-last":["error","always"]
```
Comment (anusreesubash) [inline interns-react-weekly-assignments/.eslintrc.json:25]: Please rebase to main and resolve conflicts to not have other changes in you PR.

**Review by anusreesubash — CHANGES_REQUESTED** [review]
Please rebase to main and resolve conflicts to not have other changes in you PR.

File: interns-react-weekly-assignments/src/components/Home/Aditya/Week3/components/counter/Counter.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+const Counter: React.FC = () => {
+  const [count, setCount] = useState<number>(0);
+  const handleChangeCounter = (e: React.MouseEvent<HTMLButtonElement>):void => {
+    setCount(count + (+e.currentTarget.innerText));
+  };
+
+  const changeOptions:number[] = [-1, 1, -2, 2, -5, 5, -10, 10];
+
+  return (
+    <CounterDiv>
```
Comment (choudhary-keshav) [inline interns-react-weekly-assignments/src/components/Home/Aditya/Week3/components/counter/Counter.tsx]: "CounterDiv" is not an appropriate name to use. Assign some meaningful names.

File: interns-react-weekly-assignments/src/components/Home/Aditya/Week3/components/counter/Counter.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      <h1>Counter App</h1>
+      <div className='flex'>
+        <div className='buttonDiv'>
```
Comment (choudhary-keshav) [inline interns-react-weekly-assignments/src/components/Home/Aditya/Week3/components/counter/Counter.tsx]: flex is already a common style property, do not use it for assigning classes to HTML elements.

File: interns-react-weekly-assignments/src/components/Home/Aditya/Week3/components/counter/Counter.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+          <h1 id='countValue' className='items'>
+            {count}
+          </h1>
+        </div>
+      </div>
+    </CounterDiv>
+  );
+};
+
+export default Counter;
```
Comment (choudhary-keshav) [inline interns-react-weekly-assignments/src/components/Home/Aditya/Week3/components/counter/Counter.tsx]: Specify EOL error in your .eslintrc.json under "rules" section as,
"rules": {
"eol-last": ["error", "always"]
},

File: interns-react-weekly-assignments/src/components/Home/Aditya/Week3/index.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import Status from './components/status/Status';
+
+const Week3: React.FC = () => {
+  return (
+    <Assignment3Div>
+      <div id='counter' className='relative'>
+        <Counter></Counter>
+      </div>
+      <div id='form' className='relative'>
+        <Form></Form>
```
Comment (choudhary-keshav) [inline interns-react-weekly-assignments/src/components/Home/Aditya/Week3/index.tsx]: use single tag if there is not content inside it.
Like this :- <Form />

File: interns-react-weekly-assignments/src/components/Home/Aditya/Week3/styles.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  #counter {
+    border-radius: 20px 20px 0px 0px;
+    background-color: darkolivegreen;
+  }
+  #form {
+    background-color: darkslateblue;
+  }
+  #status {
+    border-radius: 0px 0px 20px 20px;
+    background-color: midnightblue;
```
Comment (choudhary-keshav) [inline interns-react-weekly-assignments/src/components/Home/Aditya/Week3/styles.ts]: I think you're mostly using Hexadecimal format for colors

File: src/components/Home/Aditya/Week3/components/counter/Counter.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+const Counter: React.FC = () => {
+  const [count, setCount] = useState<number>(0);
+  const handleChangeCounter = (e: React.MouseEvent<HTMLButtonElement>):void => {
+    setCount(count + (+e.currentTarget.innerText));
+  };
+
+  const changeOptions:number[] = [-1, 1, -2, 2, -5, 5, -10, 10];
+
+  return (
+    <CounterDiv>
```
Comment (Harrrrry) [inline src/components/Home/Aditya/Week3/components/counter/Counter.tsx]: I recommend, CounterWrapper instead of CounterDiv

File: src/components/Home/Aditya/Week3/components/counter/Counter.tsx [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      <h1>Counter App</h1>
+      <div className='flex'>
+        <div className='buttonDiv'>
+          {changeOptions.map((value) => (
+            <button key={value} className='counterButton items' onClick={handleChangeCounter} disabled={count + value < 0}>
```
Comment (Harrrrry) [inline src/components/Home/Aditya/Week3/components/counter/Counter.tsx]: What if we create a function and put there and return Boolean
count + value < 0}

File: src/components/Home/Aditya/Week3/components/form/Form.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            type=''
+            value={profile.phone}
+            placeholder='123-456-7890'
+            pattern='[0-9]{3}-[0-9]{3}-[0-9]{4}'
+            onChange={handleChange}
+            required
+          />
+          <label htmlFor='email'>Email:</label>
+          <input name='email' type='email' value={profile.email} onChange={handleChange} required />
+          <button id='submit' type='submit'>
```
Comment (Harrrrry) [inline src/components/Home/Aditya/Week3/components/form/Form.tsx:65]: Is this submitting the form, Type submit will refresh the page.
Can you please confirm

File: src/components/Home/Aditya/Week3/components/searchUser/SearchUser.tsx [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+interface User {
+  name: string;
+  status: string;
+}
+
+const SearchUser: React.FC = () => {
+  const [search, setSearch] = useState<string>('');
+  const results: User[] = useSearch(search);
+
+  const statusChoices: string[] = ['Busy', 'Active', 'Do not disturb', 'Idle'];
```
Comment (Harrrrry) [inline src/components/Home/Aditya/Week3/components/searchUser/SearchUser.tsx]: This is duplicate, Please put in separate file

File: src/components/Home/Aditya/Week3/index.tsx [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+import Counter from './components/counter/Counter';
+import Form from './components/form/Form';
+import { Assignment3Div } from './styles';
+import Status from './components/status/Status';
+
+const Week3: React.FC = () => {
+  return (
+    <Assignment3Div>
+      <div id='counter' className='relative'>
+        <Counter></Counter>
```
Comment (Harrrrry) [inline src/components/Home/Aditya/Week3/index.tsx]: You can use `<Counter />`

File: src/components/Home/Aditya/Week3/utils/utils.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+export const removeWhitespace = (str: string): string => {
+  let currentTask = str.replace(/\s+/g, ' ');
+  if (currentTask) {
+    if (currentTask[0] === ' ') {
```
Comment (Harrrrry) [inline src/components/Home/Aditya/Week3/utils/utils.ts]: Why extra space is there in condition '  '

File: src/components/Home/Aditya/Week3/components/searchUser/SearchUser.tsx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        : user
+    );
+    localStorage.setItem('users', JSON.stringify(updatedUsers));
+  };
+
+  return (
+    <SearchUserDiv>
+      <div id='container'>
+        <h1>Search</h1>
+        <label htmlFor='search' className='font17'>
```
Comment (anusreesubash) [inline src/components/Home/Aditya/Week3/components/searchUser/SearchUser.tsx]: font17 is not a proper classname, use something else.

File: src/components/Home/Aditya/Week3/index.tsx (hunk 11 above)
Comment (anusreesubash) [reply src/components/Home/Aditya/Week3/index.tsx]: @aditya-s-19  please check across the project

File: src/components/Home/Aditya/Week3/utils/utils.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
+export const removeWhitespace = (str: string): string => {
+  let currentTask = str.replace(/\s+/g, ' ');
+  if (currentTask) {
```
Comment (anusreesubash) [inline src/components/Home/Aditya/Week3/utils/utils.ts]: if currentTask is always going to be a string do we really need this check? Try to refactor this code.

File: src/components/Home/Aditya/Week3/components/status/statusStyles.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  left: 50%;
+  transform: translate(-50%, -50%);
+  width: 400px;
+  background-color: #ffe4c4;
+  border-radius: 20px;
+  padding: 20px;
+  box-shadow:
+    inset 10px 10px 10px #87ceeb,
+    inset -10px -10px 10px #e9967a;
+  font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
```
Comment (anusreesubash) [inline src/components/Home/Aditya/Week3/components/status/statusStyles.ts:16]: check formatting

File: src/components/Home/Aditya/Week3/components/searchUser/SearchUser.tsx [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+  const statusChoices: string[] = ['Busy', 'Active', 'Do not disturb', 'Idle'];
+
+  const handleSearch = (e: React.ChangeEvent<HTMLInputElement>): void => {
+    setSearch(e.target.value);
+  };
+
+  const handleStatusChange = (e: React.ChangeEvent<HTMLSelectElement>): void => {
+    const users = JSON.parse(localStorage.getItem('users') ?? '[]');
+    const updatedUsers = users.map((user: User) =>
```
Comment (anusreesubash) [inline src/components/Home/Aditya/Week3/components/searchUser/SearchUser.tsx:18]: extract name and vale

const {name, value} = e.target

File: src/components/Home/Aditya/Week3/components/counter/Counter.tsx (hunk 7 above)
Comment (anusreesubash) [reply src/components/Home/Aditya/Week3/components/counter/Counter.tsx]: @aditya-s-19  check at all places

File: package-lock.json [hunk 17]
```diff
@@ -34,7 +34,6 @@
         "eslint-config-prettier": "^9.1.0",
```
Comment (anusreesubash) [inline package-lock.json:34]: add package-lock to gitignore file this should not be pushed

File: src/App.tsx [hunk 18]
```diff
@@ -23,7 +24,9 @@ const App: React.FC = () => {
               <Route path=':pageId' element={<Week4 />}></Route>
             </Route>
           </Route>
-          <Route path='aditya' element={<Aditya />}></Route>
+          <Route path='/aditya' element={<Aditya />}>
+            <Route path='/aditya/week3' element={<AdityaWeek3 />}></Route>
```
Comment (RitikaGarg-ITT) [inline src/App.tsx:28]: don't use your name in the other components rather you can write week3Main or something like that. Only initial component name should be of your name

File: src/components/Home/Aditya/Week3/components/newUser/NewUser.tsx [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+const NewUser: React.FC = () => {
+  const [users, setUsers] = useState<User[]>(JSON.parse(localStorage.getItem('users') ?? '[]'));
+  const [isValidUsername, setIsValidUsername] = useState<boolean>(false);
+  const [newUser, setNewUser] = useState<User>({
+    name: '',
+    status: ''
+  });
+
+  const statusChoices: string[] = ['Busy', 'Active', 'Do not disturb', 'Idle'];
```
Comment (RitikaGarg-ITT) [inline src/components/Home/Aditya/Week3/components/newUser/NewUser.tsx]: To reduce the code redundancy you can use export statusOption function which keshav made in his folder so that repetition of code will not be there

File: src/components/Home/Aditya/Week3/components/searchUser/SearchUser.tsx [hunk 20]
```diff
@@ -0,0 +1,59 @@
+import React, { useState } from 'react';
+import { SearchUserDiv } from './searchUserStyles';
+import useSearch from '../../hooks/useSearch';
+
+interface User {
+  name: string;
+  status: string;
+}
```
Comment (RitikaGarg-ITT) [inline src/components/Home/Aditya/Week3/components/searchUser/SearchUser.tsx]: make seperate folder of interfaces and there define all the interfaces

File: src/components/Home/Aditya/Week3/hooks/useSearch.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
+import { useEffect, useState } from 'react';
+import { removeWhitespace } from '../utils/utils';
+
+interface User {
+  name: string;
+  status: string;
+}
```
Comment (RitikaGarg-ITT) [inline src/components/Home/Aditya/Week3/hooks/useSearch.ts]: same for this

File: src/components/Home/Aditya/Week3/components/status/statusStyles.ts (hunk 15 above)
Comment (aditya-s-19) [reply src/components/Home/Aditya/Week3/components/status/statusStyles.ts:16]: ES lint is doing this on its own, I cannot change it

File: src/App.tsx (hunk 18 above)
Comment (aditya-s-19) [reply src/App.tsx:28]: Other will also import their weeks, so many week3 variables will be confusing

**Review by shreegovinda — APPROVED** [review]

**Review by anusreesubash — APPROVED** [review]

File: src/components/Home/Aditya/Week3/utils/utils.ts (hunk 12 above)
Comment (aditya-s-19) [reply src/components/Home/Aditya/Week3/utils/utils.ts]: It stops the string from having spaces at start
we can use trim(), but that will remove spaces at end which we don't want.
